File: internal/models/task.go
package models

import (
	"time"
)

// TaskStatus represents the current state of a task
type TaskStatus string

const (
	TaskStatusPending    TaskStatus = "pending"
	TaskStatusInProgress TaskStatus = "in_progress"
	TaskStatusCompleted  TaskStatus = "completed"
	TaskStatusCancelled  TaskStatus = "cancelled"
	TaskStatusOnHold     TaskStatus = "on_hold"
)

// TaskPriority represents the priority level of a task
type TaskPriority string

const (
	TaskPriorityLow      TaskPriority = "low"
	TaskPriorityMedium   TaskPriority = "medium"
	TaskPriorityHigh     TaskPriority = "high"
	TaskPriorityCritical TaskPriority = "critical"
)

// SortOrder represents the sort order for task listing
type SortOrder string

const (
	SortOrderCreatedAtAsc  SortOrder = "created_at_asc"
	SortOrderCreatedAtDesc SortOrder = "created_at_desc"
	SortOrderDueDateAsc    SortOrder = "due_date_asc"
	SortOrderDueDateDesc   SortOrder = "due_date_desc"
	SortOrderPriorityAsc   SortOrder = "priority_asc"
	SortOrderPriorityDesc  SortOrder = "priority_desc"
)

// Task represents a task in the system
type Task struct {
	ID          string       `json:"id"`
	Title       string       `json:"title" binding:"required,min=1,max=200"`
	Description string       `json:"description"`
	Status      TaskStatus   `json:"status"`
	Priority    TaskPriority `json:"priority"`
	Tags        []string     `json:"tags"`
	CreatedBy   string       `json:"created_by"`
	AssignedTo  *string      `json:"assigned_to,omitempty"`
	CreatedAt   time.Time    `json:"created_at"`
	UpdatedAt   time.Time    `json:"updated_at"`
	DueDate     *time.Time   `json:"due_date,omitempty"`
	CompletedAt *time.Time   `json:"completed_at,omitempty"`
}

// CreateTaskRequest represents a request to create a new task
type CreateTaskRequest struct {
	Title       string       `json:"title" binding:"required,min=1,max=200"`
	Description string       `json:"description"`
	Priority    TaskPriority `json:"priority"`
	Tags        []string     `json:"tags"`
	AssignedTo  *string      `json:"assigned_to,omitempty"`
	DueDate     *time.Time   `json:"due_date,omitempty"`
}

// UpdateTaskRequest represents a request to update an existing task
type UpdateTaskRequest struct {
	Title       *string       `json:"title,omitempty" binding:"omitempty,min=1,max=200"`
	Description *string       `json:"description,omitempty"`
	Status      *TaskStatus   `json:"status,omitempty"`
	Priority    *TaskPriority `json:"priority,omitempty"`
	Tags        []string      `json:"tags,omitempty"`
	AssignedTo  *string       `json:"assigned_to,omitempty"`
	DueDate     *time.Time    `json:"due_date,omitempty"`
}

// UpdateTaskStatusRequest represents a request to update task status
type UpdateTaskStatusRequest struct {
	Status TaskStatus `json:"status" binding:"required"`
}

// ListTasksQuery represents query parameters for listing tasks
type ListTasksQuery struct {
	PageSize   int        `form:"page_size,default=20" binding:"min=1,max=100"`
	PageToken  string     `form:"page_token"`
	Status     TaskStatus `form:"status"`
	AssignedTo string     `form:"assigned_to"`
	Tags       string     `form:"tags"`
	SortOrder  SortOrder  `form:"sort_order"`
}

// ListTasksResponse represents the response for listing tasks
type ListTasksResponse struct {
	Tasks         []Task `json:"tasks"`
	NextPageToken string `json:"next_page_token,omitempty"`
	TotalCount    int    `json:"total_count"`
}

// ErrorResponse represents an error response
type ErrorResponse struct {
	Error ErrorDetail `json:"error"`
}

// ErrorDetail contains error details
type ErrorDetail struct {
	Code    string      `json:"code"`
	Message string      `json:"message"`
	Details interface{} `json:"details,omitempty"`
}

// Validate validates the task status
func (s TaskStatus) IsValid() bool {
	switch s {
	case TaskStatusPending, TaskStatusInProgress, TaskStatusCompleted,
		TaskStatusCancelled, TaskStatusOnHold:
		return true
	default:
		return false
	}
}

// Validate validates the task priority
func (p TaskPriority) IsValid() bool {
	switch p {
	case TaskPriorityLow, TaskPriorityMedium, TaskPriorityHigh, TaskPriorityCritical:
		return true
	default:
		return false
	}
}

// GetPriorityValue returns a numeric value for priority sorting
func (p TaskPriority) GetValue() int {
	switch p {
	case TaskPriorityLow:
		return 1
	case TaskPriorityMedium:
		return 2
	case TaskPriorityHigh:
		return 3
	case TaskPriorityCritical:
		return 4
	default:
		return 2
	}
}

File: app/models/task.py
"""
Task models and DTOs.

Translated from internal/models/task.go
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class TaskStatus(str, Enum):
    """Represents the current state of a task."""

    pending = "pending"
    in_progress = "in_progress"
    completed = "completed"
    cancelled = "cancelled"
    on_hold = "on_hold"

    def is_valid(self) -> bool:
        """Check whether the status value is a known member."""
        return self.value in {member.value for member in TaskStatus}


class TaskPriority(str, Enum):
    """Represents the priority level of a task."""

    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"

    def is_valid(self) -> bool:
        """Check whether the priority value is a known member."""
        return self.value in {member.value for member in TaskPriority}

    def get_value(self) -> int:
        """Return a numeric value for priority sorting (mirrors Go GetValue)."""
        _map = {
            TaskPriority.low: 1,
            TaskPriority.medium: 2,
            TaskPriority.high: 3,
            TaskPriority.critical: 4,
        }
        return _map.get(self, 2)


class SortOrder(str, Enum):
    """Represents the sort order for task listing."""

    created_at_asc = "created_at_asc"
    created_at_desc = "created_at_desc"
    due_date_asc = "due_date_asc"
    due_date_desc = "due_date_desc"
    priority_asc = "priority_asc"
    priority_desc = "priority_desc"


# ---------------------------------------------------------------------------
# Core Task model
# ---------------------------------------------------------------------------

class Task(BaseModel):
    """Internal representation of a task (mirrors Go Task struct)."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    title: str
    description: str = ""
    status: TaskStatus
    priority: TaskPriority
    tags: list[str] = Field(default_factory=list)
    created_by: str = ""
    assigned_to: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    due_date: Optional[datetime] = None
    completed_at: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Request / Response DTOs
# ---------------------------------------------------------------------------

class CreateTaskRequest(BaseModel):
    """Request body for POST /api/v1/tasks (mirrors Go CreateTaskRequest)."""

    title: str = Field(..., min_length=1, max_length=200)
    description: str = ""
    priority: Optional[TaskPriority] = None
    tags: Optional[list[str]] = None
    assigned_to: Optional[str] = None
    due_date: Optional[datetime] = None


class UpdateTaskRequest(BaseModel):
    """Request body for PUT /api/v1/tasks/{id} (mirrors Go UpdateTaskRequest).

    All fields are optional; only provided fields are applied.
    """

    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    tags: Optional[list[str]] = None
    assigned_to: Optional[str] = None
    due_date: Optional[datetime] = None


class UpdateTaskStatusRequest(BaseModel):
    """Request body for PATCH /api/v1/tasks/{id}/status."""

    status: TaskStatus


class ListTasksQuery(BaseModel):
    """Query parameters for GET /api/v1/tasks."""

    page_size: int = Field(default=20, ge=1, le=100)
    page_token: Optional[str] = None
    status: Optional[TaskStatus] = None
    assigned_to: Optional[str] = None
    tags: Optional[str] = None
    sort_order: Optional[SortOrder] = None


class ListTasksResponse(BaseModel):
    """Response for GET /api/v1/tasks."""

    tasks: list[Task]
    next_page_token: Optional[str] = None
    total_count: int


# ---------------------------------------------------------------------------
# Error models
# ---------------------------------------------------------------------------

class ErrorDetail(BaseModel):
    """Nested error detail (mirrors Go ErrorDetail)."""

    code: str
    message: str
    details: Optional[object] = None


class ErrorResponse(BaseModel):
    """Top-level error envelope (mirrors Go ErrorResponse)."""

    error: ErrorDetail
